src/org/rust/lang/core/resolve/util/RustResolveUtil.kt

```diff
@@ -10,22 +12,48 @@ public object RustResolveUtil {
         while (current != null) {
             when (current) {
                 is RustResolveScope -> return current
-                else -> current = current.parent
+                else                -> current = current.parent
             }
         }
 
         return null
     }
 
-    fun getGlobalResolveScopeFor(elem: PsiElement): RustResolveScope {
-        var current = getResolveScopeFor(elem)
-        while (true)
-        {
-            val p = current?.let { getResolveScopeFor(it) } ?: break
-            current = p;
+    fun getCrateRootFor(elem: PsiElement): RustResolveScope? {
+        val mod = getSelfModFor(elem) ?: return null
+
+        val superMod = getSuperModFor(mod)
+        return if (superMod == null) mod else getCrateRootFor(superMod)
+    }
+
+    fun getSuperModFor(elem: RustModItem): RustModItem? {
+        val self = getSelfModFor(elem) ?: return null
+        val superInFile = self.parentOfType<RustModItem>()
+        if (superInFile != null) {
+            return superInFile
         }
 
-        return current!!;
+        val file = self.containingFile
+        val dir = self.containingFile?.containingDirectory ?: return null
+        val dirOfParent = if (self.ownsDirectory) dir.parent else dir
+        dirOfParent?.files.orEmpty()
+            .map { it.firstChild }
```

Comment:
Agree, but how exactly should I add a getter? There is no `RustFile`, only `RustFileImpl`. So should I add the getter to the `Impl` and add a `filterType<RustFileImpl>` here?

Is there any way to get files only of a specific `Language` from a `PsiDirectory`?